ssl/ech/ech_internal.c

```diff
@@ -1100,6 +1100,1025 @@ int ossl_ech_calc_confirm(SSL_CONNECTION *s, int for_hrr,
     return rv;
 }
 
+/*!
+ * Given a CH find the offsets of the session id, extensions and ECH
+ * pkt is the CH
+ * sessid_off points to offset of session_id length
+ * exts_off points to offset of extensions
+ * ech_off points to offset of ECH
+ * echtype points to the ext type of the ECH
+ * inner 1 if the ECH is marked as an inner, 0 for outer
+ * sni_off points to offset of (outer) SNI
+ * return 1 for success, other otherwise
+ *
+ * Offsets are set to zero if relevant thing not found.
+ * Offsets are returned to the type or length field in question.
+ *
+ * Note: input here is untrusted!
+ */
+int ossl_ech_get_ch_offsets(SSL_CONNECTION *s, PACKET *pkt, size_t *sessid_off,
```

Comment:
I am really unhappy about this function. This just breaks the model we follow everywhere else, and it seems unnecessary to me.

This function parses the ClientHello, and due to the fact that it is called multiple times, we can end up parsing the ClientHello four times! i.e. `tls_process_client_hello` calls `ossl_ech_get_ch_offsets` (so that's the first parse). It then calls `ossl_ech_early_decrypt` which also calls `ossl_ech_get_ch_offsets` (so that's the second parse). Then `ossl_ech_early_decrypt` calls `hpke_decrypt_encch` wich also calls `ossl_ech_get_ch_offsets` (so that's the third parse), and finally we return to `tls_process_client_hello` again and do the final, "real" parse (so that's the fourth parse).

This just doesn't make sense to me. I am sure we can avoid it.

Reply:
I'll leave this one 'till the obvious stuff below is done.

Reply:
I'm playing with this now - first cut will be to address the (valid) comment about doing the parsing over and over, by storing the values in the `OSSL_ECH_CONN` struct.  That might cause some HRR fun, but should be more or less mechanical. After that we can probably improve things more, but I'll aim to get that working first, then push that for further review.

Reply:
I just pushed a version that addresses the inefficiency part of `ossl_ech_get_ch_offsets` and that works for normal operation and HRR. That means one parsing pass on the outer CH, then one on the inner after successful ECH decryption. (And repeated for the 2nd CH in the case of HRR). If that's the right direction, we can improve it to avoid the 2nd calls to `ossl_ech_get_ch_offsets` and just use the fields in the `OSSL_ECH_CONN` struct direct, but there might be even better things to do so I've not done that yet.